File: day07.py
"""Solution to day 7 of Advent of Code"""

from get_input import get_input, line_parser
from collections import defaultdict
import re


def part1(lines):
    contained = defaultdict(set)
    for bag, contains in lines:
        for _, contain in contains:
            contained[contain].add(bag)
    queue = ['shiny gold']
    seen = set()
    while queue:
        bag = queue.pop()
        if bag in seen:
            continue
        seen.add(bag)
        queue.extend(contained[bag])
    return len(seen) - 1


def part2(lines):
    contained = dict(lines)
    total = 0
    queue = list(contained['shiny gold'])
    while queue:
        count, bag_type = queue.pop()
        total += count
        for inner_count, inner_type in contained[bag_type]:
            queue.append((inner_count * count, inner_type))
    return total


def parse(line):
    bag = ' '.join(line.split(' ')[:2])
    contains = []
    # 1 muted blue bag
    for (count, bag_type) in re.findall(r'(\d+) (\w+ \w+) bags?', line):
        contains.append((int(count), bag_type))
    return (bag, tuple(contains))


test_text = """
light red bags contain 1 bright white bag, 2 muted yellow bags.
dark orange bags contain 3 bright white bags, 4 muted yellow bags.
bright white bags contain 1 shiny gold bag.
muted yellow bags contain 2 shiny gold bags, 9 faded blue bags.
shiny gold bags contain 1 dark olive bag, 2 vibrant plum bags.
dark olive bags contain 3 faded blue bags, 4 dotted black bags.
vibrant plum bags contain 5 faded blue bags, 6 dotted black bags.
faded blue bags contain no other bags.
dotted black bags contain no other bags.
"""

test_text_part_2 = """
shiny gold bags contain 2 dark red bags.
dark red bags contain 2 dark orange bags.
dark orange bags contain 2 dark yellow bags.
dark yellow bags contain 2 dark green bags.
dark green bags contain 2 dark blue bags.
dark blue bags contain 2 dark violet bags.
dark violet bags contain no other bags.
"""

if __name__ == "__main__":
    assert part1(line_parser(test_text, parse=parse)) == 4
    assert part2(line_parser(test_text_part_2, parse=parse)) == 126
    LINES = line_parser(get_input(day=7, year=2020), parse=parse)
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day07.rs
use common::get_input;
use std::collections::{VecDeque, HashSet, HashMap};
use regex::Regex;

type Bag = String;
type Input = HashMap<Bag, HashSet<(usize, Bag)>>;

fn part1(bags: &Input) -> usize {
    let mut container_map = HashMap::new();
    for (bag, containers) in bags {
        for (_, container) in containers {
            container_map.entry(container).or_insert(HashSet::new()).insert(bag);

        }
    }
    let mut queue: VecDeque<String> = VecDeque::new();
    let mut seen: HashSet<String> = HashSet::new();
    queue.push_back("shiny gold".to_string());
    while let Some(bag_type) = queue.pop_front() {
        if seen.contains(&bag_type) {
            continue;
        }
        seen.insert(bag_type.to_string());
        for container in container_map.get(&bag_type.to_string()).unwrap_or(&HashSet::new()).iter() {
            queue.push_back(container.to_string());
        }
    }
    seen.len() - 1
}

fn part2(bags: &Input) -> usize {
    let mut total = 0;
    let mut queue = VecDeque::new();
    queue.push_back((1, "shiny gold"));
    while let Some((count, bag_type)) = queue.pop_front() {
        total += count;
        for (inner_count, inner_type) in bags.get(bag_type).unwrap().iter() {
            queue.push_back((inner_count * count, inner_type));
        }
    }
    total - 1
}

fn parse(text: String) -> Input {
    let mut map = HashMap::new();
    let bag_regex = Regex::new(r"^(\w+ \w+)").unwrap();
    let contains_regex = Regex::new(r"(\d+) (\w+ \w+)").unwrap();
    for line in text.trim().split('\n') {
        let bag = bag_regex.find(line).unwrap().as_str();
        map.insert(bag.to_string(), HashSet::new());
        for contains in contains_regex.captures_iter(line) {
            let count = contains[1].parse::<usize>().unwrap();
            let container = contains[2].to_string();
            map.get_mut(&bag.to_string()).unwrap().insert((count, container));
        }
    }
    map
}

fn main() {
    let input = parse(get_input(07, 2020));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
